### PR TITLE
Updates to address changes in upcoming MultiQC 1.0

### PR DESCRIPTION
**Pull request by vladsavelyev**
MultiQC 1.0 to be released this week, and there are API changes that better to be addressed in modules (some of them seem to be breaking): https://github.com/ewels/MultiQC/blob/master/CHANGELOG.md#v10dev
- using wrapper `self.add_section` instead of explicit `self.sections.append`
- search mechanism:
  - using flattened search keys (`"bcbio/metrics"`)
  - `get_filelist()` needs to be called to apply `config.sp` changes in module's `__init__` (well, populating `config` anywhere outside the documented locations http://multiqc.info/docs/#configuring-multiqc looks like a hack, so no surprise it got broken - probably need to coordinate with @ewels to make this properly)
- using config options to set numeric multiplier (`config.read_count_multiplier`, `read_count_prefix` and `read_count_desc`)
- new number formatting (`{:.1f}%` > `{:,.1f}`)

### CODE REVIEW

**Comment by lpantano** [comment]
Hi @vladsaveliev 

this sounds great! do you know why is failing travis? is like this new feature of `update_dict` is not in the master branch of multiqc. Is this happening somewhere else?

thanks!

**Comment by ewels** [comment]
Yes, it's a function that he wrote in the PR that he has open on MultiQC at the moment :) Will merge later this morning then we can restart the test..

**Comment by vladsavelyev** [comment]
That's right, sorry about using it before the PR got merged. I was hoping that Phil will come and say that we gotta remove that dirty hack for good :)

File: multiqc_bcbio/bcbio.py
Comment (ewels) [inline multiqc_bcbio/bcbio.py]: If `mirna_stats.mirs` is just the returned plot data (looks like it is?) then please use `plot = mirna_stats.mirs` instead. Not much difference currently, but this new syntax is to enable better consistent formatting in the future for help text. Basically `content` = `description`+`help`+`plot`.

File: multiqc_bcbio/bcbio.py
Comment (ewels) [inline multiqc_bcbio/bcbio.py:177]: Need to also have `'suffix': '%'` in the config, else we'll lose the % symbol.

File: multiqc_bcbio/bcbio.py
Comment (ewels) [inline multiqc_bcbio/bcbio.py]: Same as above with `suffix` and removing `%` from the format string. Apologies, need to improve the docs on this I think.

File: multiqc_bcbio/bcbio.py
Comment (vladsavelyev) [reply multiqc_bcbio/bcbio.py:177]: It's there, in one line with `'min': 0, 'max': 100, 'suffix': '%',`

File: multiqc_bcbio/bcbio.py
Comment (vladsavelyev) [reply multiqc_bcbio/bcbio.py]: My fault - the idea is clear, but I didn't spot all places where it need to be changed...

**Comment by ewels** [comment]
Yes, the config modification won't work now as it'll be too late - the plugin code runs as a module, by the time it's executed the files have already been discovered.

Instead, need to have a function tied to the `config_loaded ` hook which runs after the config files have been loaded (before the command line config). This is early enough that it'll then be used when looking for files and stuff.

File: multiqc_bcbio/bcbio.py
Comment (ewels) [reply multiqc_bcbio/bcbio.py:177]: ah, sorry - didn't see it 👍

File: multiqc_bcbio/bcbio.py
Comment (ewels) [reply multiqc_bcbio/bcbio.py]: The docs for making tables (not General Stats) don't mention the `suffix` key either though, will add that now. Also for `{:n}x` below.

**Comment by vladsavelyev** [comment]
It actually works now after I added an explicit call of `get_filelist()` right after changing `config.sp`. But I figure that adding a hook is a more correct way (despite more wordy). However I was thinking about the idea of adding another yaml config location, like module and plugin specific, and put it in the priority order on top of `config_defaults.yaml` but before any kind of user configs.

**Comment by ewels** [comment]
Yup, but calling `get_filelist()` a second time will almost double the execution time. Probably not a problem for smaller reports, but can take quite a long time with more files.

Module / plugin config would be nice - you mean add a new plugin hook there? Or is there another nicer way to specify extra config files to load?

**Comment by vladsavelyev** [comment]
The core has its own defaults built-in configuration YAML file https://github.com/ewels/MultiQC/blob/master/multiqc/utils/config_defaults.yaml, plus it separately has the defaults for the `sp` section at `search_patterns.yaml`.
All user configs are placed on top of that in order described here http://multiqc.info/docs/#configuring-multiqc
But none of those ways works to preconfigure an external module like `bcbio`: we can't add to core's `search_patterns.yaml` and `config_defaults.yaml` because the module is not in the core. And we can't use any of those system-wide and user personal ways to configure and add search patterns. That's why we end up doing that in python code in `__init__.py`.

So I'm thinking it would be nice to have a `config_defaults.yaml` or `search_patterns.yaml` somewhere inside of https://github.com/MultiQC/MultiQC_bcbio/tree/master/multiqc_bcbio, so we could put there the bcbio-specific search patterns, `fn_clean_exts`, and `table_columns_visible` that we redefine.

**Comment by ewels** [comment]
Yeah sure - I understand the point :) But MultiQC needs to know that this file exists and where to load it from. That's the bit I was asking about: how do you suggest telling core MultiQC that a plugin or module wants an additional config file to be loaded?

**Comment by vladsavelyev** [comment]
Ah, sorry, you are thinking too far ahead of me...

In `multiqc` main script, load the modules before `report.get_filelist()`, and in-between check if the module has something in its root dir? Might be not reliable, and also requires to specify YAML in module's `setup.py` and manifest.in to make sure it's included it into the package. Also, there are other options used before modules are loaded, like `config.title` and `config.module_order`, that a module developer might want to redefine, and that will lead to reordering more stuff in the main script, not good. 

I guess using the `config_loaded` hook as you suggested is the most natural solution :)

**Comment by ewels** [comment]
Yeah, I thought of doing that - looping through the modules at config load time. But the setuptools hooks directly reference a function, not a package. So the only way to really "see" them is by actually running them, which isn't ideal here.

I think that the best (only?) way to do this is to have a new setuptools hook that plugins / modules can tie into with a function that returns a config dict. This should be pretty easy to implement and use.. I'll add this now.

**Comment by ewels** [comment]
Hi both,

I've added the new config hook (see above commit) and it seems to work. I've just done similar updates to our [MultiQC_NGI](https://github.com/ewels/MultiQC_NGI) plugin.

Specifically, [this hook](https://github.com/ewels/MultiQC_NGI/blob/master/setup.py#L51) calls [this function](https://github.com/ewels/MultiQC_NGI/blob/master/multiqc_ngi/multiqc_ngi.py#L26-L36) which adds to the default config. Seems to work really well!

Phil

**Comment by ewels** [comment]
It's a bit late now, but I just wrote a few docs about what plugin authors need to update for this release (mostly using this PR as a reference!). http://multiqc.info/docs/#v1.0-updates

**Comment by vladsavelyev** [comment]
Thanks a lot Phil. I pushed the changes here to update the config using the new hook. Travis is passing. @lpantano, looks good to merge now, I hope.

**Comment by ewels** [comment]
Looks good to me! I'll leave it to @lpantano to merge if that's ok as I don't run bcbio myself.

**Review by ewels — APPROVED** [review]

**Comment by lpantano** [comment]
You are awesome! thanks!

**Pull request merged**
